packages/vite/src/node/plugins/resolve.ts

```diff
@@ -875,14 +875,30 @@ export async function tryOptimizedResolve(
   const metadata = depsOptimizer.metadata
 
   const depInfo = optimizedDepInfoFromId(metadata, id)
-  if (depInfo) {
-    return depsOptimizer.getOptimizedDepId(depInfo)
+  if (!importer) {
+    // no importer. try our best to find an optimized dep
+    if (depInfo) {
+      return depsOptimizer.getOptimizedDepId(depInfo)
+    }
+    return
   }
 
-  if (!importer) return
-
-  // further check if id is imported by nested dependency
+  // resolvedSrc is what the importer wants
+  // we can return a optimizedDep only if optimizedDep.src === resolvedSrc
+  // https://github.com/vitejs/vite/pull/11290
   let resolvedSrc: string | undefined
+  try {
+    // this may throw errors if unable to resolve, e.g. aliased id
+    resolvedSrc = normalizePath(resolveFrom(id, path.dirname(importer)))
+  } catch {
+    // this is best-effort only so swallow errors
+  }
+  if (!resolvedSrc) return
+
+  // check if the found optimizedDep comes from the resolvedSrc
+  if (depInfo && depInfo.src === resolvedSrc) {
```

Comment:
I check and return`depInfo` in the `importer exists` case, too. But I don't know if it is necessary.

What is the difference between matching `optimizedDepInfoFromId(metadata, id)` and matching `metadata.depInfoList`?

Reply:
Yeah this part shouldn't be necessary so we can make it lazy again.

> What is the difference between matching `optimizedDepInfoFromId(metadata, id)` and matching `metadata.depInfoList`? Do we need to do both here? @bluwy

They're both kinda the same thing, except `optimizedDepInfoFromId` accesses the objects via a key so it's faster. You can think it as equivalent of `metadata.depInfoList.find(info => info.id === id)`.